### PR TITLE
Add searx.locale.get_engine_locale in addition to searx.utils.match_language; improve qwant engine

### LINKED ISSUE #1640: Add ability to disable or enable safesearch on Qwant web & image search
**Is your feature request related to a problem? Please describe.**
Searxng has so far not implemented safesearch on it's Qwant module yet but this can be easily done using the 's' parameter.

**Describe the solution you'd like**
For any search on it's web or image section, simply attach &s=0 to disable safesearch, &s=1 to enable moderate safesearch or &s=2 to enable strict safesearch.

This is with strict safesearch: 

https://www.qwant.com/?s=2&q=dead+lion&t=images&s=2

This is with no safesearch:

https://www.qwant.com/?s=2&q=dead+lion&t=images&s=0

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by return42**
## What does this PR do?

Various improvements to the qwant engine

- `[mod] decouple qwant's categories from SearXNG's categories`

  By using new property `qwant_categ:` the category of qwant is no longer bound to the category of SearXNG.

- `[mod] add locale.get_engine_locale to get predictable results` ...

   A function that is used where match_languages cannot be used or gives worse results. Needed by the next patch ..

- `[fix] qwant - API error::locale must be one` ..

  The request function should not request a language (aka locale) that is not supported by qwant. Select a locale like zh-TW ends in qwant's API error: ..

- `[mod] qwant - add safesearch option`

  see https://github.com/searxng/searxng/issues/1640

## Related issues

Closes: https://github.com/searxng/searxng/issues/1640

### CODE REVIEW

**Comment by tiekoetter** [comment]
@return42 
I have a few questions / some comments:
1. Regarding 84d163530d97830c716964f9d13e2bdff7fccba0: What is the benefit of decoupling the Qwant engine specifically from the categories?
2. Regarding 564058fb857db5cb90d1de1349942cbc8e0aaaf8: I understand the need to map the locales between what SearXNG uses and what the specific engine uses. But I can't reproduce what you have mentioned should result in an API error. Searching `!qw time :zh-TW` using the current master branch does not result in an API error. Could further explain this please?
3. Regarding 6fe821143b07935dae300bf0b23ee7249b0fb42e: Works as it should 👍

I think you have a typo in the commit message of 564058fb857db5cb90d1de1349942cbc8e0aaaf8 `The quant's _fetch_supported_languages` and `8aka locales)`.

**Comment by return42** [comment]
>  1. Regarding [84d1635](https://github.com/searxng/searxng/commit/84d163530d97830c716964f9d13e2bdff7fccba0): What is the benefit of decoupling the Qwant engine specifically from the categories?

By example, we have: 

https://github.com/searxng/searxng/blob/43c545910b161731c3bc6791bdfa227c64e4489a/searx/settings.yml#L1183-L1186

and you wan't to remove qwant-images from category `images` and place it in `mycateg`:

```yaml 
 - name: qwant images 
   engine: qwant 
   shortcut: qwi 
   categories: mycateg 
```

then the qwant-images engine wan't work any longer when you use `!mycateg`:

https://github.com/searxng/searxng/blob/43c545910b161731c3bc6791bdfa227c64e4489a/searx/engines/qwant.py#L54-L55

> 2. Regarding [564058f](https://github.com/searxng/searxng/commit/564058fb857db5cb90d1de1349942cbc8e0aaaf8): I understand the need to map the locales between what SearXNG uses and what the specific engine uses. 

OK

> But I can't reproduce what you have mentioned should result in an API error. Searching `!qw time :zh-TW` using the current master branch does not result in an API error. Could further explain this please?

Thanks for pointing out; now I realize that the language issue is only in the qwant-news engine

    !qwn paris :zh-TW

I have to modify the patch .. I will have a look the next days ..

> 3. Regarding [6fe8211](https://github.com/searxng/searxng/commit/6fe821143b07935dae300bf0b23ee7249b0fb42e): Works as it should +1

:+1:

> I think you have a typo in the commit message of [564058f](https://github.com/searxng/searxng/commit/564058fb857db5cb90d1de1349942cbc8e0aaaf8) `The quant's _fetch_supported_languages` and `8aka locales)`.

Oops . thanks for pointing out .. I will fix it.

File: searx/engines/qwant.py [hunk 1]
```diff
@@ -238,19 +243,43 @@ def response(resp):
...
+    q_locales = q_initial_props.get('locales')
+    q_valid_locales = []
+
+    for country, v in q_locales.items():
+        for lang in v['langs']:
+            _locale = "{lang}_{country}".format(lang=lang, country=country)
+
+            if qwant_categ == 'news' and _locale.lower() not in qwant_news_locales:
+                # qwant-news does not support all locales from qwant-web:
+                continue
```
Comment (return42) [inline searx/engines/qwant.py:262]: > But I can't reproduce what you have mentioned should result in an API error. Searching !qw time :zh-TW using the current master branch does not result in an API error. Could further explain this please?

@tiekoetter  I fixed this: its only the qwant-news engine that do not supports some qwant locales / with the fix from above we can see that qwant-news support 24 languages (aka locales) while other qwant engines support 39 locales:

```
$ ./manage pyenv.cmd ./searxng_extra/update/update_languages.py
...
fetched 39 languages from engine qwant
fetched 39 languages from engine qwant images
fetched 24 languages from engine qwant news
fetched 39 languages from engine qwant videos
```

**Comment by tiekoetter** [comment]
Regarding 793b570e42bc00029f4b0b047d3efbb2583b6ffb:
I have some minor corrections regarding the commit message. I also must say this is the most detailed commit message which I have ever seen!

``` 
  A. When user select a language the results should be optimized according to
     the selected language.

  B. When user select a language and a territory the results should be
     optimized with first priority on terrirtory and second on language.
```
to
```
  A. When a user selects a language the results should be optimized according to
     the selected language.

  B. When user selects a language and a territory the results should be
     optimized with first priority on territory and second on language.
```
----

```
Example: (Assumption A.)

  User selects region 'zh-TW' what should end in zh_HK
```
to
```
Example: (Assumption A.)

  A user selects region 'zh-TW' which should end in zh_HK
```
----

```
Example: (Assumption A.)

  User selects only the language 'zh' what should end in CN
```
to
```
Example: (Assumption A.)

  A user selects only the language 'zh' which should end in CN
```
----

```
Example: (Assumption B.)

  User selects region 'fr-BE' what should end in nl-BE

hint:
  priority should be first on territory the user selected.  If user
  prefer 'fr' he will select 'fr' without a region tag.
```
to
```
Example: (Assumption B.)

  A user selects region 'fr-BE' which should end in nl-BE

hint:
  priority should be on the territory the user selected.  If the user
  prefers 'fr' he will select 'fr' without a region tag.
```
----

```
Example: (Assumption A.)

  User selects only the language 'fr' what should end in fr_CA
```
to
```
Example: (Assumption A.)

  A user selects only the language 'fr' which should end in fr_CA
```
----

```
The difference of priotity on terrirtory is best shown with a engine that
supports the follwoing locales:
```
to
```
The difference in priority on the territory is best shown with a engine that
supports the following locales:
```
----

```
Example: (Assumption A.)

   User selects only a language
```
to
```
Example: (Assumption A.)

   A user selects only a language

```
----

```
hint: the engine supports fr_FR and fr_CA sine no terrirtory is given, fr_FR
takes priority ..
```
to 
```
hint: the engine supports fr_FR and fr_CA since no territory is given, fr_FR
takes priority ..
```
----

```
Example: (Assumption B.)

  User selects region 'fr-BE' what should end in nl-BE
```
to
```
Example: (Assumption B.)

  A user selects region 'fr-BE' which should end in nl-BE
```
----

```
Lokks both function return the same value, but match_language depends oon the
order of the dictionary (what is not predictable):
```
to
```
Looks like both functions return the same value, but match_language depends on the
order of the dictionary (which is not predictable):
```

**Review by tiekoetter — APPROVED** [review]
LGTM
If https://github.com/searxng/searxng/pull/1652#issuecomment-1214216527 is addressed

**Comment by return42** [comment]
> If https://github.com/searxng/searxng/pull/1652#issuecomment-1214216527 is addressed

I fixed the commit message .. thanks for reading with care :+1: :+1: 

> I also must say this is the most detailed commit message which I have ever seen!

:laughing: 

TL;DR;

I prefer git comments to document a change process.  Most people prefer short git comments and left the detailed discussions of the changes in the issues.

I think the collaboration tools of git hosters are valuable for community collaboration.  However, the insights gained from the exchange should be additional and directly readable in the commit messages of the changes.  

That's why I take the liberty to re-worden the commits and to take passages from the posts to the PR.

The requirement or its implementation (such as a function or class description) belongs in my opinion directly in the source code, as close as possible to the place of implementation. In python code, we can use comments or better doc-strings.

----

I think the git hosts like gitlab, githost and many others (e.g. gitea instances) are excellent tools to promote collaboration in a community.  However, in most cases an account is required and personalization of activities takes place.  What is done with the personalized data is usually not transparent and at least the question remains, what motivation drives a top dog like github.

From a privacy point of view, one could, for example, be critical of our hoster (github).  However, the choice of hoster is usually more dependent on the question of which hoster the majority of a community will be found on.
